.github/workflows/ci.yml

```diff
@@ -149,11 +149,6 @@ jobs:
           RUST_BACKTRACE: 1
         run: nix develop -L --no-update-lock-file --command just impure-tests
 
-      - name: "CLI Integration Tests"
-        env:
-          RUST_BACKTRACE: 1
-        run: nix develop -L --no-update-lock-file --command just integ-tests
-
```

Comment:
> - cli-dev: runs them "locally" on two different Runner systems
> - nix-build-bats-tests: runs them "remotely" on four different builder systems

in addition to that, `cli-dev` is meant to test the development flox, i.e. flox built with cargo on a dev's machine, while `nix-build-bats-tests` tests the nix build artifact, i.e. the released package.
IIRC the former tries to ensure that running integration tests during development is consistent with the released package.

I think @garbas can add on that

Reply:
Can you describe what circumstances the two types of job might have different behaviour?

Is that worth the additional runtime for every change?

Reply:
In the past we've had instances where well-intentioned changes to the testing infrastructure (`flox-cli-tests`, `flox-tests`, other things in the `bats` test suite) broke the local development workflow, so we verify that the local development workflow continues to work.

We _also_ run the test suite against a Nix-built `flox` since that binary may be slightly different from the one that's built with a `cargo build`.

Reply:
I agree that it's not ideal that we're running the test suite twice.

Reply:
A saving of 7mins wall time and 22mins billable time seems too good to ignore.

I don't really understand how the outcomes could be different, or why `flox-cli-tests` is so much more involved, so it would be great if you have any examples.

Reply:
I do recommend cutting back on the OSX jobs in some way. One possibility is to either have the development tests to do only builds. Or to do builds + subset of tests (as this PR does). Or apply this cutting back only to the OSX category.

Reply:
Looking at it from another angle; it feels problematic that there are three different ways to run the tests and that we have to support them all in parallel. What could be done to converge them?

Reply:
Two different ways (nix, vs cargo) seem rather necessary as we dont want to put nix builds of the entire flox into the dev cycle.
Problems with that

a) are that nix builds dont have a compile cache like cargo has. this can be partially mitigated and i fact we already try to do so already by compiling crate dependencies separately so only the flox crates need to be rebuilt usually. incremental builds of flox with cargo are still much faster as they often finish before nix has even evaluated the flox package (that seems like a tangential issue, but #1303 suggests there's not much room for improvement).

b) dev tooling mostly expects cargo, or inevitably invokes cargo anyway, meaning we still have to support that mode anyway


What's the third mode you see?
-------

I wonder if we can run these "devshell sanity tests" outside of the main CI though, e.g. as a cron job or only if test files changed?

Reply:
That explanation helps, thanks!

> What's the third mode you see?

There's two variations of nix run tests: `.#flox-tests` and `.#flox-cli-tests`. The former appears to only be referenced by the README now, so could it be removed in favour of `nix develop … just integ-tests`?

> I wonder if we can run these "devshell sanity tests" outside of the main CI though, e.g. as a cron job and/or in ci only if test files changed?

I suspect it would be annoying to maintain a list of files that may _potentially_ affect the tests or to find out that the tests were broken after a delay (which is kinda the situation we have with the functional tests right now).

Running just one of the integration test files and renaming the CI job to make it clearer why we're running it might be a good compromise though. What do you think? Does it matter which test file it is, e.g. should it do some project setup?

After this we should figure out what impact it has on #1566.

Reply:
`flox-tests` is the functional tests, just FYI, so that can't be removed.

Reply:
> `flox-tests` is the functional tests, just FYI, so that can't be removed.

Can you link to where it's used?

Reply:
https://github.com/flox/flox/blob/93a11d728b796ab4ec41d23f52a3deb85d952a07/Justfile#L78